lib/logging/src/main/java/org/sagebionetworks/kinesis/AwsKinesisRecordIterator.java

```diff
@@ -0,0 +1,103 @@
+package org.sagebionetworks.kinesis;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+
+import com.amazonaws.services.kinesisfirehose.model.Record;
+
+public class AwsKinesisRecordIterator implements Iterator<AwsKinesisRecord> {
+	
+	private List<? extends AwsKinesisLogRecord> records;
+	private AwsKinesisLogRecordSerializer recordSerializer;
+	private int recordSizeLimit;
+	private int currentIndex;
+
+	public AwsKinesisRecordIterator(List<? extends AwsKinesisLogRecord> records, AwsKinesisLogRecordSerializer recordSerializer, int recordSizeLimit) {
+		this.records = records;
+		this.recordSerializer = recordSerializer;
+		this.recordSizeLimit = recordSizeLimit;
+		this.currentIndex = 0;
+	}
+
+	@Override
+	public boolean hasNext() {
+		return currentIndex < records.size();
+	}
+
+	@Override
+	public AwsKinesisRecord next() {
+		// Consume all the records up to the size limit
+		AwsKinesisRecordBuilder builder = new AwsKinesisRecordBuilder(recordSerializer, recordSizeLimit);
+
+		for (; currentIndex < records.size(); currentIndex++) {
+			boolean added = builder.putRecord(records.get(currentIndex));
+			if (currentIndex == 0 && !added) {
+				throw new IllegalStateException("A single record cannot exceed the limit of " + recordSizeLimit + " bytes.");
+			}
+			if (!added) {
+				break;
+			}
+		}
+
+		return builder.build();
+	}
+	
+	/**
+	 * Helper to build a single kinesis firehose record from a set of {@link AwsKinesisLogRecord}. Will pack a serialized JSON records each separated by a newline
+	 * 
+	 */
+	private static final class AwsKinesisRecordBuilder {
+		
+		private AwsKinesisLogRecordSerializer serializer;
+		private int recordSizeLimit;
+		private ByteArrayOutputStream byteArrayOutputStream;
+
+		public AwsKinesisRecordBuilder(AwsKinesisLogRecordSerializer serializer, int recordSizeLimit) {
+			this.serializer = serializer;
+			this.recordSizeLimit = recordSizeLimit;
+			this.byteArrayOutputStream = new ByteArrayOutputStream();
+		}
+		
+		/**
+		 * Try to add the given record to this builder if it does exceed the record limit of kinesis firehose
+		 * 
+		 * @param record The record to add
+		 * @return True if the record was added, false if the kinesis firehose records size limit is exceeded
+		 */
+		public boolean putRecord(AwsKinesisLogRecord record) {
+			if (byteArrayOutputStream.size() >= recordSizeLimit) {
+				return false;
+			}
+			
+			byte[] jsonBytes = serializer.toBytes(record);
+			
+			// If we are over the limit don't add the record and return
+			if (byteArrayOutputStream.size() + jsonBytes.length + AwsKinesisFirehoseConstants.NEW_LINE_BYTES.length > recordSizeLimit) {
+				return false;
+			}
+
+			try {
+				byteArrayOutputStream.write(jsonBytes);
```

Comment:
can the jsonbytes contain a new line?

Reply:
I tested this and it seems to work fine. Will add this case into the integration test.